hokusai/services/command_runner.py

```diff
@@ -1,71 +1,177 @@
     self.ecr = ECR()
+    self.pod_name = self._name()
+    self.container_name = self.pod_name
 
-  def run(self, tag_or_digest, cmd, tty=None, env=(), constraint=()):
-    if os.environ.get('USER') is not None:
-      # The regex used for the validation of name is '[a-z0-9]([-a-z0-9]*[a-z0-9])?'
-      user = re.sub("[^0-9a-z]+", "-", os.environ.get('USER').lower())
-      uuid = "%s-%s" % (user, k8s_uuid())
-    else:
-      uuid = k8s_uuid()
+  def _name(self):
+    ''' generate name for pod and container '''
+    name = '-'.join(
+      filter(
+        None,
+        [
+          f'{config.project_name}-hokusai-run',
+          user(),
+          k8s_uuid()
+        ]
+      )
+    )
+    return name
 
-    name = "%s-hokusai-run-%s" % (config.project_name, uuid)
-    separator = "@" if ":" in tag_or_digest else ":"
-    image_name = "%s%s%s" % (self.ecr.project_repo, separator, tag_or_digest)
+  def _image_name(self, tag_or_digest):
+    ''' generate docker image name '''
+    separator = '@' if ':' in tag_or_digest else ':'
+    image_name = f'{self.ecr.project_repo}{separator}{tag_or_digest}'
+    return image_name
 
-    container = {
-      "args": cmd.split(' '),
-      "name": name,
-      "image": image_name,
-      "imagePullPolicy": "Always",
-      'envFrom': [{'configMapRef': {'name': "%s-environment" % config.project_name}}]
-    }
+  def _set_env(self, container_spec, env):
+    ''' set env field of given container spec '''
+    spec = copy.deepcopy(container_spec)
+    spec['env'] = []
+    for var in env:
+      validate_key_value(var)
+      split = var.split('=', 1)
+      spec['env'].append(
+        {'name': split[0], 'value': split[1]}
+      )
+    return spec
 
-    run_tty = tty if tty is not None else config.run_tty
-    if run_tty:
-      container.update({
-        "stdin": True,
-        "stdinOnce": True,
-        "tty": True
-      })
+  def _set_envfrom(self, container_spec):
+    ''' set envFrom field of given container spec '''
+    spec = copy.deepcopy(container_spec)
+    spec.update(
+      {
+        'envFrom': [
+          {
+            'configMapRef': {
+              'name': f'{config.project_name}-environment'
+            }
+          },
+          {
+            'secretRef': {
+              'name': f'{config.project_name}',
+              'optional': True
+            }
+          }
+        ]
+      }
+    )
+    return spec
 
-    if env:
-      container['env'] = []
-      for s in env:
-        if '=' not in s:
-          raise HokusaiError("Error: environment variables must be of the form 'KEY=VALUE'")
-        split = s.split('=', 1)
-        container['env'].append({'name': split[0], 'value': split[1]})
-
-    spec = { "containers": [container] }
-    constraints = constraint or config.run_constraints
-    if constraints:
+  def _set_constraint(self, containers_spec, constraint):
+    ''' set nodeSelector field of given containers spec '''
+    spec = copy.deepcopy(containers_spec)
```

Comment:
Again not convinced you need to deepcopy the arg if you are just setting it based on the return value of this function in line 117